### PR TITLE
clarify that neither Reliable Size nor error code is a reliable signal

### LINKED ISSUE #33: Reliable Size can be greater than that specified by the application
We correctly remind that QUIC stacks might over-deliver, stating that: _As described in [Section 3.2](https://rfc-editor.org/rfc/rfc9000#section-3.2) of [[RFC9000](https://quicwg.org/reliable-stream-reset/draft-ietf-quic-reliable-stream-reset.html#RFC9000)], it MAY deliver data beyond that offset to the application._

But we do not explicitly state that QUIC stacks may over-deliver Reliable Size; i.e., the value might be greater than what is provided by the application, though I think that has always been our intent. I think it would be better to state that so that people using QUIC do not make the incorrect assumption.

Similarly, we should state that the error code might not get delivered to the peer application.

### PR DESCRIPTION
**Pull request by kazuho**
Closes #33.

### CODE REVIEW

File: draft-ietf-quic-reliable-stream-reset.md [hunk 1]
```diff
@@ -146,8 +146,11 @@ up to that byte offset are lost, the initiator MUST retransmit this data, as
 described in ({{Section 13.3 of RFC9000}}). Data sent beyond that byte offset
 SHOULD NOT be retransmitted.
 
-As described in {{Section 3.2 of RFC9000}}, it MAY deliver data beyond that
-offset to the application.
+As described in {{Section 3.2 of RFC9000}}, QUIC stacks MAY deliver data beyond
+that offset to the receiving application. An sending QUIC stack MAY choose a
```
Comment (marten-seemann) [inline draft-ietf-quic-reliable-stream-reset.md:150]: I don't understand this sentence. This seems to be up to the API contract between the sender's QUIC stack and the application. I'd consider it a very weird API if the application requests a reset with reliable size X, but the QUIC stack might sends a RESET_STREAM_AT with reliable size Y, but that's none of our business.

All that the protocols cares about is that a RESET_STREAM_AT was sent with offset Y.

File: draft-ietf-quic-reliable-stream-reset.md [hunk 2]
```diff
@@ -146,8 +146,11 @@ up to that byte offset are lost, the initiator MUST retransmit this data, as
...
+Reliable Size that is greater than the offset supplied by the application.
+Applications MUST NOT assume that the error code or the offset traverses to their
+peers.
```
Comment (marten-seemann) [inline draft-ietf-quic-reliable-stream-reset.md:153]: See https://github.com/quicwg/reliable-stream-reset/issues/33#issuecomment-1765653993, this is part of the API contract, and I don't think we should specify anything here.

File: draft-ietf-quic-reliable-stream-reset.md (hunk 1 above)
Comment (kazuho) [reply draft-ietf-quic-reliable-stream-reset.md:150]: > I'd consider it a very weird API if the application requests a reset with reliable size X, but the QUIC stack might sends a RESET_STREAM_AT with reliable size Y, but that's none of our business.

FWIW, this actually helps reduce sender state. When application requests a Reliable Reset, [POC of quicly](https://github.com/h2o/quicly/pull/550/commits/2326f9572b8d01687f2af398d8b1cf016aee2a0c#diff-c63dc437bd0da3e0f0e1ac685a5bc07e769a0064ea96ab7159e08e4efeeaf474R6511-R6513) chooses `max(reliable_size, size_inflight)` as the value of Final Size. The only additional state is a flag indicating if the value has to be communicated using FIN or if the value has to be sent as "Final Size" and also "Reliable Size" using a RESET_STREAM_AT frame.

File: draft-ietf-quic-reliable-stream-reset.md (hunk 2 above)
Comment (kazuho) [reply draft-ietf-quic-reliable-stream-reset.md:153]: I disagree, let's continue discussing this on #33.

**Comment by kazuho** [comment]
@marten-seemann Following the discussion on #33, I've shortened the text, avoiding the RFC 2119 verbs as we are merely pointing out what is written down in RFC 9000. PTAL.

File: draft-ietf-quic-reliable-stream-reset.md [hunk 3]
```diff
@@ -146,8 +146,9 @@ up to that byte offset are lost, the initiator MUST retransmit this data, as
 described in ({{Section 13.3 of RFC9000}}). Data sent beyond that byte offset
 SHOULD NOT be retransmitted.
 
-As described in {{Section 3.2 of RFC9000}}, it MAY deliver data beyond that
-offset to the application.
+As described in {{Section 3 of RFC9000}}, stream reset signals are not guaranteed
+to traverse to the receiver. QUIC stacks might deliver data beyond the specified
+offset to the receiving application.
```
Comment (marten-seemann) [inline draft-ietf-quic-reliable-stream-reset.md:151]:
```suggestion
As described in {{Section 3.2 of RFC9000}}, a stream reset signal might be
suppressed or withheld, and the same applies to a stream reset signal carried in
a RESET_STREAM_AT frame. Similary, the Reliable Size of the RESET_STREAM_AT
frame doesn't prevent the QUIC stacks from delivering more data beyond the
specified offset to the receiving application.
```

**Review by marten-seemann — COMMENTED** [review]
I reworded the paragraph a bit. @kazuho wdyt?

File: draft-ietf-quic-reliable-stream-reset.md [hunk 4]
```diff
@@ -146,8 +146,11 @@ up to that byte offset are lost, the initiator MUST retransmit this data, as
 described in ({{Section 13.3 of RFC9000}}). Data sent beyond that byte offset
 SHOULD NOT be retransmitted.
 
-As described in {{Section 3.2 of RFC9000}}, it MAY deliver data beyond that
-offset to the application.
+As described in {{Section 3.2 of RFC9000}}, a stream reset signal might be
+suppressed or withheld, and the same applies to a stream reset signal carried in
+a RESET_STREAM_AT frame. Similary, the Reliable Size of the RESET_STREAM_AT
+frame doesn't prevent the QUIC stacks from delivering more data beyond the
```
Comment (kazuho) [inline draft-ietf-quic-reliable-stream-reset.md:152]:
```suggestion
frame does not prevent the QUIC stacks from delivering data beyond the
```
Maybe:
* doesn't -> does not
* more data beyond -> data beyond

**Comment by kazuho** [comment]
@marten-seemann Thanks this looks better!

File: draft-ietf-quic-reliable-stream-reset.md [hunk 5]
```diff
@@ -146,8 +146,11 @@ up to that byte offset are lost, the initiator MUST retransmit this data, as
 described in ({{Section 13.3 of RFC9000}}). Data sent beyond that byte offset
 SHOULD NOT be retransmitted.
 
-As described in {{Section 3.2 of RFC9000}}, it MAY deliver data beyond that
-offset to the application.
+As described in {{Section 3.2 of RFC9000}}, a stream reset signal might be
+suppressed or withheld, and the same applies to a stream reset signal carried in
+a RESET_STREAM_AT frame. Similary, the Reliable Size of the RESET_STREAM_AT
+frame does not prevent the QUIC stacks from delivering data beyond the
```
Comment (marten-seemann) [inline draft-ietf-quic-reliable-stream-reset.md:152]:
```suggestion
frame does not prevent a QUIC stack from delivering data beyond the
```

**Review by marten-seemann — APPROVED** [review]

**Pull request merged**